go/arrow/compute/internal/kernel.go

```diff
@@ -0,0 +1,508 @@
+func (fsbLikeMatcher) Equals(other TypeMatcher) bool {
+	_, ok := other.(fsbLikeMatcher)
+	return ok
+}
+
+// Integer returns a TypeMatcher which will match any integral type like int8 or uint16
+func Integer() TypeMatcher { return integerMatcher{} }
+
+// BinaryLike returns a TypeMatcher that will match Binary or String
+func BinaryLike() TypeMatcher { return binaryLikeMatcher{} }
+
+// LargeBinaryLike returns a TypeMatcher which will match LargeBinary or LargeString
+func LargeBinaryLike() TypeMatcher { return largeBinaryLikeMatcher{} }
+
+// FixedSizeBinaryLike returns a TypeMatcher that will match FixedSizeBinary
+// or Decimal128/256
+func FixedSizeBinaryLike() TypeMatcher { return fsbLikeMatcher{} }
+
+type primitiveMatcher struct{}
+
+func (primitiveMatcher) String() string                  { return "primitive" }
+func (primitiveMatcher) Matches(typ arrow.DataType) bool { return arrow.IsPrimitive(typ.ID()) }
+func (primitiveMatcher) Equals(other TypeMatcher) bool {
+	_, ok := other.(primitiveMatcher)
+	return ok
+}
+
+// Primitive returns a TypeMatcher that will match any type that arrow.IsPrimitive
+// returns true for.
+func Primitive() TypeMatcher { return primitiveMatcher{} }
+
+// InputKind is an enum representing the type of Input matching
+// that will be done. Either accepting any type, an exact specific type
+// or using a TypeMatcher.
+type InputKind int8
+
+const (
+	InputAny InputKind = iota
+	InputExact
+	InputUseMatcher
+)
+
+type InputType struct {
+	Kind    InputKind
+	Type    arrow.DataType
+	Matcher TypeMatcher
+}
+
+func NewExactInput(dt arrow.DataType) InputType { return InputType{Kind: InputExact, Type: dt} }
+func NewMatchedInput(match TypeMatcher) InputType {
+	return InputType{Kind: InputUseMatcher, Matcher: match}
+}
+func NewIDInput(id arrow.Type) InputType { return NewMatchedInput(SameTypeID(id)) }
+
+func (it InputType) String() string {
+	switch it.Kind {
+	case InputAny:
+		return "any"
+	case InputUseMatcher:
+		return it.Matcher.String()
+	case InputExact:
+		return it.Type.String()
+	}
+	return ""
+}
+
+func (it *InputType) Equals(other *InputType) bool {
+	if it == other {
+		return true
+	}
+
+	if it.Kind != other.Kind {
+		return false
+	}
+
+	switch it.Kind {
+	case InputAny:
+		return true
+	case InputExact:
+		return arrow.TypeEqual(it.Type, other.Type)
+	case InputUseMatcher:
+		return it.Matcher.Equals(other.Matcher)
+	default:
+		return false
+	}
+}
+
+func (it InputType) Hash() uint64 {
+	var h maphash.Hash
+
+	h.SetSeed(hashSeed)
+	result := HashCombine(h.Sum64(), uint64(it.Kind))
+	switch it.Kind {
+	case InputExact:
+		result = HashCombine(result, arrow.HashType(hashSeed, it.Type))
+	}
+	return result
+}
+
+func (it InputType) Matches(dt arrow.DataType) bool {
+	switch it.Kind {
+	case InputExact:
+		return arrow.TypeEqual(it.Type, dt)
+	case InputUseMatcher:
+		return it.Matcher.Matches(dt)
+	default:
+		return true
+	}
```

Comment:
Wouldn't you want `default` to error or panic, and break out Any as its own branch?

Reply:
I was matching the logic in https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernel.cc#L334 

Though it might make sense to break Any as it's own branch and put a `debug.Assert` for the default

Reply:
I guess I have the same nit there :) though it's not a big deal.

Reply:
added a `debug.Assert` and a separate branch for `InputAny`